src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java

```diff
@@ -1197,4 +1197,31 @@ public InputStream getFileContent(BitbucketSCMFile file) throws IOException, Int
         }
         return content;
     }
+
+    @Restricted(ProtectedExternally.class)
+    protected CloseableHttpResponse executeMethod(HttpRequestBase httpMethod)
+            throws IOException {
+        CloseableHttpClient client = getHttpClient(httpMethod);
+        CloseableHttpResponse response = client.execute(httpMethod, context);
+        while (response.getStatusLine().getStatusCode() == TOO_MANY_REQUESTS_HTTP_STATUS) {
+            response.close();
+            httpMethod.releaseConnection();
+
+            /*
+             * TODO: When bitbucket starts supporting rate limit expiration time, remove 5
+             * sec wait and put code to wait till expiration time is over. It should also
+             * fix the wait for ever loop.
+             */
+            LOGGER.fine("Bitbucket server API rate limit reached, sleeping for 5 sec then retry...");
```

Comment:
Maybe better to print the given time, since it is configurable.
So e.g. `"... sleeping for %d ms ....", REQUEST_WAIT_TIME_DEFAULT)`

Reply:
Completely agree, handled.